src/api/libcellml/annotator.h

```diff
@@ -0,0 +1,1054 @@
+     * @param index The position of an item within the list of items with the given @p id to retrieve.
+     *
+     * @return A @c ComponentPtr.
+     */
+    ComponentPtr componentRef(const std::string &id, size_t index);
+
+    /**
+     * @brief Retrieve a @c ResetPtr whose test_value has the given @p id.
+     *
+     *        If an item with the id is not found, or has another type, the
+     *        @c nullptr is returned.
+     *
+     * @param id String representing the id of the item to retrieve.
+     *
+     * @return A @c ResetPtr.
+     */
+    ResetPtr testValue(const std::string &id);
+
+    /**
+     * @brief From a list of items in the stored model with the given @p id string,
+     *        this function returns a @c ResetPtr in the @p index position, if it exists.
+     *
+     *  A @c nullptr will be returned if:
+     *      - no item with the given @p id exists in the stored model;
+     *      - the given @p index is beyond the range [0, \#duplicateCount(id));
+     *      - the item stored at the @p index is not a @c ResetPtr; or
+     *      - the annotator is not built; 
+     *
+     * @param id A @c std::string representing the id of the item to retrieve.
+     * @param index The position of an item within the list of items with the given @p id to retrieve.
+     *
+     * @return A @c ResetPtr.
+     */
+    ResetPtr testValue(const std::string &id, size_t index);
+
+    /**
+     * @brief Retrieve a @c ResetPtr whose reset_value has the given @p id.
+     *
+     *  If an item with the id is not found, or has another type, the
+     *  @c nullptr is returned.
+     * 
+     *  The annotator index must be built using Annotator::buildModelIndex(ModelPtr model)
+     *  before this function can be called successfully.
+     *
+     * @param id String representing the id of the item to retrieve.
+     *
+     * @return A @c ResetPtr.
+     */
+    ResetPtr resetValue(const std::string &id);
+
+    /**
+     * @brief From a list of items in the stored model with the given @p id string,
+     *        this function returns a @c ResetPtr in the @p index position, if it exists.
+     *
+     *  A @c nullptr will be returned if:
+     *      - no item with the given @p id exists in the stored model;
+     *      - the given @p index is beyond the range [0, \#duplicateCount(id));
+     *      - the item stored at the @p index is not a @c ResetPtr; or
+     *      - the annotator is not built; 
+     *
+     * @param id A @c std::string representing the id of the item to retrieve.
+     * @param index The position of an item within the list of items with the given @p id to retrieve.
+     *
+     * @return A @c ResetPtr.
+     */
+    ResetPtr resetValue(const std::string &id, size_t index);
+
+    /**
+     * @brief Assign an id string to every item in the model which is already 
+     * stored in this annotator, excluding its MathML items.
+     * 
```

Comment:
Gah!! You told me to take that away!! I used to only assign a new id to things whose id was missing or invalid.  But, from our discussion, you said anyone wanting to assign ids automagically was prepared to nuke it all ... :(

Reply:
hmmm...that doesn't sound like something I would say :) is there still a print method that will add ids to things? maybe that is the one that will do what I am thinking here...

but at the moment I'm struggling to think why it would ever be ok to ignore ids that have already been set.... I can see a case for wanting to specifically nuke them all and start from scratch, but not that it would be the only method to make sure everything has an id. It would be a fairly typical use-case to want to simply make sure that an existing model has ids everywhere so that, for example, a user can annotate the model without having to edit the model.

Reply:
Haha, It's easy enough to change that back, but I'll wait until Hugh has finished his stuff.

Reply:
Actually, nothing needs changing. These functions all have an `something.empty()` test in them, so existing ids are not changed? I've changed the documentation so it's clearer.  Is that enough or did you find something that was being changed?

Reply:
8d4307d2